cedar-lean/Cedar/Thm/Validation/Validator.lean

```diff
@@ -0,0 +1,373 @@
+/-
+ Copyright Cedar Contributors
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-/
+
+import Cedar.Spec
+import Cedar.Data
+import Cedar.Validation
+import Cedar.Thm.Validation.Typechecker
+import Cedar.Thm.Data
+
+namespace Cedar.Thm
+
+open Cedar.Data
+open Cedar.Spec
+open Cedar.Validation
+
+/-- For a single expression, evaluates to a boolean value (or appropriate error) -/
+def EvaluatesToBool (expr : Expr) (request : Request) (entities : Entities) : Prop :=
+  ∃ (b : Bool), EvaluatesTo expr request entities b
+
+/-- Every policy as an expression evaluates to a boolean value or appropriate error -/
+def AllEvaluateToBool (policies : Policies) (request : Request) (entities : Entities) : Prop :=
+  ∀ policy ∈ policies, EvaluatesToBool policy.toExpr request entities
+
+def RequestAndEntitiesMatchSchema (schema : Schema) (request : Request) (entities : Entities) :Prop :=
+  ∀ env ∈ schema.toEnvironments,
+  RequestAndEntitiesMatchEnvironment env request entities
+
+def SubstituteActionPreservesEvaluation (expr : Expr) (request : Request) (entities : Entities) : Prop :=
+  evaluate (substituteAction request.action expr) request entities = evaluate expr request entities
```

Comment:
Just noticed that you're only using this predicate in hypotheses below. Is there a reason you don't use it in the main lemma statement?

Reply:
Just following the structure of the typechecker proofs -- no real reason to do it this way, I guess.